File: ios-shell/Sources/CredentialStore.swift
import CryptoKit
import Foundation

/// A saved Braze workspace credential profile. `externalId` is authoritative for
/// `changeUser` — native owns the demo user identity.
struct CredentialProfile: Codable, Equatable {
  var id: String
  var name: String
  var apiKey: String
  var endpoint: String
  var externalId: String
  /// Explicit web source override. Leave empty for the generated runtime source.
  var webURL: String

  init(
    id: String = UUID().uuidString, name: String, apiKey: String, endpoint: String,
    externalId: String, webURL: String
  ) {
    self.id = id
    self.name = name
    self.apiKey = apiKey
    self.endpoint = endpoint
    self.externalId = externalId
    self.webURL = webURL
  }

  // Tolerate profiles saved before `webURL` existed (defaults to "").
  init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    id = try c.decode(String.self, forKey: .id)
    name = try c.decode(String.self, forKey: .name)
    apiKey = try c.decode(String.self, forKey: .apiKey)
    endpoint = try c.decode(String.self, forKey: .endpoint)
    externalId = try c.decode(String.self, forKey: .externalId)
    webURL = Self.normalizeWebURL((try? c.decode(String.self, forKey: .webURL)) ?? "")
  }

  static func normalizeWebURL(_ value: String) -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return "" }
    guard let url = URL(string: trimmed), url.scheme == "http" else { return "" }
    let host = url.host ?? ""
    return ["localhost", "127.0.0.1"].contains(host) ? trimmed : ""
  }
}

/// The generated, pack-owned workspace seed compiled through the ignored
/// `Config.swift`. Its stored marker is a one-way fingerprint, never a second
/// copy of the SDK credentials.
struct GeneratedCredentialSeed: Equatable {
  let packId: String
  let packName: String
  let configHash: String
  let apiKey: String
  let endpoint: String
  let externalId: String

  static var current: GeneratedCredentialSeed {
    GeneratedCredentialSeed(
      packId: Config.demoPackId,
      packName: Config.demoPackName,
      configHash: Config.demoConfigHash,
      apiKey: Config.brazeAPIKey,
      endpoint: Config.brazeEndpoint,
      externalId: Config.demoExternalId)
  }

  var profileId: String {
    "generated:\(normalizedPackId)"
  }

  var profile: CredentialProfile? {
    let key = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
    let sdkEndpoint = endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !key.isEmpty, !sdkEndpoint.isEmpty else { return nil }
    return CredentialProfile(
      id: profileId,
      name: packName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        ? "Generated workspace"
        : packName.trimmingCharacters(in: .whitespacesAndNewlines),
      apiKey: key,
      endpoint: sdkEndpoint,
      externalId: externalId.trimmingCharacters(in: .whitespacesAndNewlines),
      webURL: "")
  }

  /// Changes when the generated workspace context changes, while revealing no
  /// API key or endpoint in UserDefaults diagnostics or exported state.
  var fingerprint: String {
    let parts = [
      "ios-generated-credential-seed/v1",
      normalizedPackId,
      configHash.trimmingCharacters(in: .whitespacesAndNewlines),
      apiKey.trimmingCharacters(in: .whitespacesAndNewlines),
      endpoint.trimmingCharacters(in: .whitespacesAndNewlines),
      externalId.trimmingCharacters(in: .whitespacesAndNewlines),
    ]
    let digest = SHA256.hash(data: Data(parts.joined(separator: "\u{0}").utf8))
    return digest.map { String(format: "%02x", $0) }.joined()
  }

  /// Safe runtime proof that the selected SDK credentials belong to this
  /// generated pack context. The plaintext key and endpoint never leave native.
  func sdkCredentialContextFingerprint(profile: CredentialProfile?) -> String {
    guard let profile else { return "" }
    let key = profile.apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
    let sdkEndpoint = profile.endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !key.isEmpty, !sdkEndpoint.isEmpty else { return "" }
    let parts = [
      "ios-sdk-credential-context/v1",
      normalizedPackId,
      configHash.trimmingCharacters(in: .whitespacesAndNewlines),
      key,
      sdkEndpoint,
    ]
    let digest = SHA256.hash(data: Data(parts.joined(separator: "\u{0}").utf8))
    return digest.map { String(format: "%02x", $0) }.joined()
  }

  private var normalizedPackId: String {
    let value = packId.trimmingCharacters(in: .whitespacesAndNewlines)
    return value.isEmpty ? "default" : value
  }
}

/// UserDefaults-backed list of credential profiles + the active selection. This is
/// the canonical credential store (the Swift SDK needs creds at init). The web
/// setup screen reads/writes it through the bridge. Survives app relaunches.
final class CredentialStore {
  static let shared = CredentialStore()

  private let profilesKey = "braze.demo.profiles"
  private let activeKey = "braze.demo.activeProfileId"
  private let generatedSeedFingerprintKey = "braze.demo.generatedSeedFingerprint"
  private let defaults: UserDefaults

  init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  var activeSdkCredentialContextFingerprint: String {
    GeneratedCredentialSeed.current.sdkCredentialContextFingerprint(profile: activeProfile)
  }

  var profiles: [CredentialProfile] {
    get {
      guard let data = defaults.data(forKey: profilesKey),
        let list = try? JSONDecoder().decode([CredentialProfile].self, from: data)
      else { return [] }
      return list
    }
    set {
      defaults.set(try? JSONEncoder().encode(newValue), forKey: profilesKey)
    }
  }

  var activeProfileId: String? {
    get { defaults.string(forKey: activeKey) }
    set {
      if let newValue {
        defaults.set(newValue, forKey: activeKey)
      } else {
        defaults.removeObject(forKey: activeKey)
      }
    }
  }

  var activeProfile: CredentialProfile? {
    guard let id = activeProfileId else { return nil }
    return profiles.first { $0.id == id }
  }

  /// Insert or update a profile (matched by id) and make it active.
  @discardableResult
  func save(_ profile: CredentialProfile) -> CredentialProfile {
    var profile = profile
    profile.webURL = CredentialProfile.normalizeWebURL(profile.webURL)
    var list = profiles
    if let idx = list.firstIndex(where: { $0.id == profile.id }) {
      list[idx] = profile
    } else {
      list.append(profile)
    }
    profiles = list
    activeProfileId = profile.id
    return profile
  }

  func select(id: String) {
    guard profiles.contains(where: { $0.id == id }) else { return }
    activeProfileId = id
  }

  /// Reconcile the compiled/generated workspace without erasing user-created
  /// profiles. A changed generated context takes over predictably. A changed
  /// context with no complete generated credentials clears the implicit active
  /// selection, so the current runtime cannot silently use a prior workspace;
  /// the operator may then explicitly select a saved profile.
  ///
  /// Returns true when the active workspace changed or was cleared.
  @discardableResult
  func reconcileGeneratedSeed(_ seed: GeneratedCredentialSeed = .current) -> Bool {
    let fingerprint = seed.fingerprint
    let previousFingerprint = defaults.string(forKey: generatedSeedFingerprintKey)
    var list = profiles
    let generatedProfile = seed.profile

    if let generatedProfile {
      let existingIndex =
        list.firstIndex(where: { $0.id == generatedProfile.id })
        ?? list.firstIndex(where: {
          $0.name == Self.legacyGeneratedSeedName
            && $0.apiKey == generatedProfile.apiKey
            && $0.endpoint == generatedProfile.endpoint
        })
      if let existingIndex {
        list[existingIndex] = generatedProfile
      } else {
        list.append(generatedProfile)
      }
      profiles = list
    }

    let activeId = activeProfileId
    let activeStillExists = activeId.map { id in list.contains(where: { $0.id == id }) } ?? false
    let generatedContextChanged = previousFingerprint != fingerprint
    var activeWorkspaceChanged = false

    if generatedContextChanged {
      activeProfileId = generatedProfile?.id
      activeWorkspaceChanged = true
    } else if !activeStillExists, let generatedProfile {
      activeProfileId = generatedProfile.id
      activeWorkspaceChanged = true
    } else if !activeStillExists {
      activeProfileId = nil
    }

    defaults.set(fingerprint, forKey: generatedSeedFingerprintKey)
    return activeWorkspaceChanged
  }

  /// JSON for export (web setup screen offers download/copy).
  func exportJSON() -> String {
    guard let data = try? JSONEncoder().encode(profiles),
      let str = String(data: data, encoding: .utf8)
    else { return "[]" }
    return str
  }

  func importJSON(_ json: String) {
    guard let data = json.data(using: .utf8),
      let list = try? JSONDecoder().decode([CredentialProfile].self, from: data)
    else { return }
    profiles = list
    if activeProfileId.map({ id in list.contains(where: { $0.id == id }) }) != true {
      activeProfileId = list.first?.id
    }
  }

  private static let legacyGeneratedSeedName = "Seed (Config.swift)"
}

File: android-shell/app/src/main/java/com/braze/demoshell/CredentialStore.kt
package com.braze.demoshell

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest
import java.util.UUID

data class CredentialProfile(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val apiKey: String,
    val endpoint: String,
    val externalId: String,
    val firstName: String = "",
    val lastName: String = "",
    val initials: String = "",
    val homeLocation: String = "",
    val homeAddress: String = "",
    val loyaltyTier: String = "",
    val loyaltyPoints: Int? = null,
    val favoriteCategories: List<String> = emptyList(),
    val webURL: String = "",
) {
    fun toJson(active: Boolean = false): JSONObject {
        val json = JSONObject()
            .put("id", id)
            .put("name", name)
            .put("apiKey", apiKey)
            .put("endpoint", endpoint)
            .put("externalId", externalId)
            .put("firstName", firstName)
            .put("lastName", lastName)
            .put("initials", initials)
            .put("homeLocation", homeLocation)
            .put("homeAddress", homeAddress)
            .put("loyaltyTier", loyaltyTier)
            .put("webURL", webURL)
            .put("active", active)
        loyaltyPoints?.let { json.put("loyaltyPoints", it) }
        json.put("favoriteCategories", JSONArray().apply {
            favoriteCategories.forEach { put(it) }
        })
        return json
    }

    companion object {
        private fun stringList(json: JSONObject, key: String): List<String> {
            val array = json.optJSONArray(key) ?: return emptyList()
            return buildList {
                for (index in 0 until array.length()) {
                    val value = array.optString(index)
                    if (value.isNotBlank()) add(value)
                }
            }
        }

        fun fromJson(json: JSONObject): CredentialProfile =
            CredentialProfile(
                id = json.optString("id").ifBlank { UUID.randomUUID().toString() },
                name = json.optString("name"),
                apiKey = json.optString("apiKey"),
                endpoint = json.optString("endpoint"),
                externalId = json.optString("externalId", CredentialStore.DEFAULT_EXTERNAL_ID),
                firstName = json.optString("firstName"),
                lastName = json.optString("lastName"),
                initials = json.optString("initials"),
                homeLocation = json.optString("homeLocation"),
                homeAddress = json.optString("homeAddress"),
                loyaltyTier = json.optString("loyaltyTier"),
                loyaltyPoints = if (json.has("loyaltyPoints")) json.optInt("loyaltyPoints") else null,
                favoriteCategories = stringList(json, "favoriteCategories"),
                webURL = normalizeWebUrl(json.optString("webURL")),
            )

        private fun normalizeWebUrl(value: String): String =
            value.trim().let { trimmed ->
                when {
                    trimmed.isBlank() -> ""
                    trimmed.startsWith("file:///android_asset/lumo/index.html") -> ""
                    trimmed == CredentialStore.DEFAULT_WEB_URL -> ""
                    isAllowedLocalWebUrl(trimmed) -> trimmed
                    else -> ""
                }
            }

        fun isAllowedLocalWebUrl(value: String): Boolean {
            val trimmed = value.trim()
            if (trimmed == CredentialStore.DEFAULT_WEB_URL) return true
            return Regex("^http://(localhost|127\\.0\\.0\\.1|10\\.0\\.2\\.2)(:\\d+)?(/.*)?$").matches(trimmed)
        }
    }
}

data class GeneratedCredentialSeed(
    val packId: String,
    val packName: String,
    val profileName: String,
    val configHash: String,
    val apiKey: String,
    val endpoint: String,
    val externalId: String,
) {
    val profileId: String
        get() = "generated:${normalizedPackId()}"

    fun profile(): CredentialProfile? {
        val key = apiKey.trim()
        val sdkEndpoint = endpoint.trim()
        if (key.isBlank() || sdkEndpoint.isBlank()) return null
        return CredentialProfile(
            id = profileId,
            name = profileName.trim().ifBlank { packName.trim().ifBlank { "Generated workspace" } },
            apiKey = key,
            endpoint = sdkEndpoint,
            externalId = externalId.trim().ifBlank { CredentialStore.DEFAULT_EXTERNAL_ID },
            webURL = "",
        )
    }

    /** One-way generated-context marker; SharedPreferences never receives another plaintext key copy. */
    fun fingerprint(): String = sha256(
        listOf(
            "android-generated-credential-seed/v1",
            normalizedPackId(),
            configHash.trim(),
            apiKey.trim(),
            endpoint.trim(),
            externalId.trim(),
        ),
    )

    /** Safe runtime proof that the active SDK credentials belong to this generated pack context. */
    fun sdkCredentialContextFingerprint(profile: CredentialProfile?): String {
        if (profile == null || profile.apiKey.isBlank() || profile.endpoint.isBlank()) return ""
        return sha256(
            listOf(
                SDK_CREDENTIAL_CONTEXT_PROTOCOL,
                normalizedPackId(),
                configHash.trim(),
                profile.apiKey.trim(),
                profile.endpoint.trim(),
            ),
        )
    }

    private fun normalizedPackId(): String = packId.trim().ifBlank { "default" }

    companion object {
        const val SDK_CREDENTIAL_CONTEXT_PROTOCOL = "android-sdk-credential-context/v1"

        fun current(): GeneratedCredentialSeed = GeneratedCredentialSeed(
            packId = BuildConfig.DEMO_PACK_ID,
            packName = BuildConfig.DEMO_PACK_NAME,
            profileName = BuildConfig.DEMO_PROFILE_NAME,
            configHash = BuildConfig.DEMO_CONFIG_HASH,
            apiKey = BuildConfig.BRAZE_API_KEY,
            endpoint = BuildConfig.BRAZE_ENDPOINT,
            externalId = BuildConfig.DEMO_EXTERNAL_ID,
        )

        private fun sha256(parts: List<String>): String =
            MessageDigest.getInstance("SHA-256")
                .digest(parts.joinToString("\u0000").toByteArray(Charsets.UTF_8))
                .joinToString("") { byte -> "%02x".format(byte) }
    }
}

data class CredentialSeedReconciliation(
    val profiles: List<CredentialProfile>,
    val activeProfileId: String?,
    val fingerprint: String,
    val contextChanged: Boolean,
)

object CredentialSeedPolicy {
    fun reconcile(
        profiles: List<CredentialProfile>,
        activeProfileId: String?,
        previousFingerprint: String?,
        seed: GeneratedCredentialSeed,
    ): CredentialSeedReconciliation {
        val fingerprint = seed.fingerprint()
        val generatedProfile = seed.profile()
        val list = profiles.toMutableList()
        if (generatedProfile != null) {
            val existingIndex = list.indexOfFirst { it.id == generatedProfile.id }.takeIf { it >= 0 }
                ?: list.indexOfFirst {
                    it.name == CredentialStore.LEGACY_SEED_NAME &&
                        it.apiKey == generatedProfile.apiKey &&
                        it.endpoint == generatedProfile.endpoint
                }.takeIf { it >= 0 }
            if (existingIndex != null) list[existingIndex] = generatedProfile else list.add(generatedProfile)
        }

        val contextChanged = previousFingerprint != fingerprint
        val activeStillExists = activeProfileId != null && list.any { it.id == activeProfileId }
        val nextActiveId = when {
            contextChanged -> generatedProfile?.id
            !activeStillExists -> generatedProfile?.id
            else -> activeProfileId
        }
        return CredentialSeedReconciliation(list, nextActiveId, fingerprint, contextChanged)
    }
}

class CredentialStore private constructor(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    var activeProfileId: String?
        get() = prefs.getString(ACTIVE_KEY, null)
        private set(value) {
            prefs.edit().putString(ACTIVE_KEY, value).apply()
        }

    val profiles: List<CredentialProfile>
        get() {
            val raw = prefs.getString(PROFILES_KEY, "[]") ?: "[]"
            val array = runCatching { JSONArray(raw) }.getOrDefault(JSONArray())
            return buildList {
                for (index in 0 until array.length()) {
                    val item = array.optJSONObject(index) ?: continue
                    runCatching { add(CredentialProfile.fromJson(item)) }
                }
            }
        }

    val activeProfile: CredentialProfile?
        get() {
            val list = profiles
            val id = activeProfileId ?: return null
            return list.firstOrNull { it.id == id }
        }

    /**
     * Returns true when the generated seed took over the active profile, i.e. the demo pack,
     * config hash, or seed identity changed. Callers use this to drop any remembered runtime
     * identity so a new pack always starts on its own seed user.
     */
    fun seedIfEmpty(): Boolean {
        val outcome = CredentialSeedPolicy.reconcile(
            profiles = profiles,
            activeProfileId = activeProfileId,
            previousFingerprint = prefs.getString(SEED_MARKER_KEY, null),
            seed = GeneratedCredentialSeed.current(),
        )
        persist(outcome.profiles)
        activeProfileId = outcome.activeProfileId
        val markerEditor = prefs.edit().putString(SEED_MARKER_KEY, outcome.fingerprint)
        if (outcome.contextChanged) markerEditor.putBoolean(SEED_CONTEXT_CHANGE_PENDING_KEY, true)
        markerEditor.apply()
        return outcome.contextChanged
    }

    fun consumeGeneratedSeedContextChange(): Boolean {
        val pending = prefs.getBoolean(SEED_CONTEXT_CHANGE_PENDING_KEY, false)
        if (pending) prefs.edit().remove(SEED_CONTEXT_CHANGE_PENDING_KEY).apply()
        return pending
    }

    fun activeSdkCredentialContextFingerprint(): String =
        GeneratedCredentialSeed.current().sdkCredentialContextFingerprint(activeProfile)

    fun save(profile: CredentialProfile): CredentialProfile {
        val list = profiles.toMutableList()
        val index = list.indexOfFirst { it.id == profile.id }
        if (index >= 0) {
            list[index] = profile
        } else {
            list.add(profile)
        }
        persist(list)
        activeProfileId = profile.id
        return profile
    }

    fun select(id: String): Boolean {
        if (profiles.none { it.id == id }) return false
        activeProfileId = id
        return true
    }

    fun profilesPayload(): JSONArray {
        val activeId = activeProfileId
        return JSONArray().apply {
            profiles.forEach { profile ->
                put(profile.toJson(active = profile.id == activeId))
            }
        }
    }

    private fun persist(list: List<CredentialProfile>) {
        val array = JSONArray()
        list.forEach { array.put(it.toJson()) }
        prefs.edit().putString(PROFILES_KEY, array.toString()).apply()
    }

    companion object {
        const val DEFAULT_EXTERNAL_ID = "lumo-demo-user"
        const val DEFAULT_WEB_URL = "file:///android_asset/demo/index.html"

        private const val PREFS = "braze.demo.credentials"
        private const val PROFILES_KEY = "braze.demo.profiles"
        private const val ACTIVE_KEY = "braze.demo.activeProfileId"
        private const val SEED_MARKER_KEY = "braze.demo.seedMarker"
        private const val SEED_CONTEXT_CHANGE_PENDING_KEY = "braze.demo.seedContextChangePending"
        internal const val LEGACY_SEED_NAME = "Seed (local.properties)"

        @Volatile
        private var instance: CredentialStore? = null

        fun get(context: Context): CredentialStore =
            instance ?: synchronized(this) {
                instance ?: CredentialStore(context).also { instance = it }
            }
    }
}
